snakemake/deployment/spack.py

```diff
@@ -0,0 +1,232 @@
+from enum import Enum
+import threading
+import shutil
+from .base import EnvBase
+
+from snakemake.exceptions import CreateSpackEnvironmentException, WorkflowError
+from snakemake.logging import logger
+from snakemake.common import strip_prefix, ON_WINDOWS
+from snakemake import utils
+from snakemake.deployment import singularity, containerize
+from snakemake.io import git_content
+from snakemake.sourcecache import LocalGitFile, LocalSourceFile, infer_source_file
+
+
+class SpackCleanupMode(Enum):
+    tarballs = "tarballs"
+    cache = "cache"
+
+    def __str__(self):
+        return self.value
+
+
+class Env(EnvBase):
+
+    """Spack environment from a given specification file."""
+
+    def __init__(self, env_file, workflow, env_dir=None, cleanup=None):
+
+        self.file = infer_source_file(env_file)
+        self.workflow = workflow
+
+        self._env_dir = env_dir or workflow.persistence.spack_env_path
+        self._hash = None
+        self._content_hash = None
+        self._content = None
+        self._path = None
+        self._archive_file = None
+        self._cleanup = cleanup
+
+    @property
+    def _env_archive_dir(self):
+        return self.workflow.persistence.spack_env_archive_path
+
+    @property
+    def hash(self):
+        if self._hash is None:
+            md5hash = hashlib.md5()
+            # Include the absolute path of the target env dir into the hash.
+            # By this, moving the working directory around automatically
+            # invalidates all environments. This is necessary, because binaries
+            # in spack environments can contain hardcoded absolute RPATHs.
+            env_dir = os.path.realpath(self._env_dir)
+            md5hash.update(env_dir.encode())
+            md5hash.update(self.content)
+            self._hash = md5hash.hexdigest()
+        return self._hash
+
+    @property
+    def path(self):
+        """Path to directory of the spack environment.
+
+        First tries full hash, if it does not exist, (8-prefix) is used
+        as default.
+        """
+        env_dir = self._env_dir
+        get_path = lambda h: os.path.join(env_dir, h)
+        hash_candidates = [self.hash[:8], self.hash]
+        exists = [os.path.exists(get_path(h)) for h in hash_candidates]
+        if exists[1] or (not exists[0]):
+            # full hash exists or fallback hash does not exist: use full hash
+            return get_path(hash_candidates[1])
+        # use fallback hash
+        return get_path(hash_candidates[0])
+
+    def create_archive(self):
+        """
+        Create self-contained archive of environment.
+
+        This won't be possible with spack until an environment can be created
+        that uses a copy instead of link (now possible) but ALSO can be
+        relocated. Currently, the spack.yaml files are the best reproducers.
+        """
+        raise WorkflowError("spack does not support environment archive.")
+
+    def create(self, dryrun=False):
```

Comment:
Perhaps this method can share code with the conda implementation as well?

Reply:
Sure I can give a shot - I didn't originally because there are enough differences that trying to separate out into shared code would increase complexity for the developer by quite a bit to jump back and forth.